### PR TITLE
Fix race between member join and listener registration/deregistration

### PR DESCRIPTION
**Pull request by mdogan**
Registering/deregistering a listener is racy while a new member is joining
cluster.
Solution is to use `InvocationUtil.invokeOnStableClusterSerial()` to submit operations
and check memberlist version while registering/deregistering the listener. A similar
mechanism used by dynamic data-structure config registration...

Additionally, EventService supports both post-join operation and pre-join operation.
Prejoin operation ensures that data structure operations cannot be executed on newly joining
member before event listener is registered. Postjoin operation is required newly joining member
to register listeners on the cluster, such as proxy service listener.

Fixes https://github.com/hazelcast/hazelcast/issues/11490
Fixes https://github.com/hazelcast/hazelcast/issues/11115

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/AbstractRegistrationOperation.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    public final void run() throws Exception {
+        runInternal();
+        checkMemberListVersion();
+    }
+
+    protected abstract void runInternal() throws Exception;
+
+    private void checkMemberListVersion() {
+        ClusterService clusterService = getNodeEngine().getClusterService();
+        if (clusterService.isMaster()) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/AbstractRegistrationOperation.java]: Shouldn't we also test for `memberListVersion > -1` here? On a 3.9 member working in 3.8 cluster version, `readData` will not deserialize any member list version.

Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/operations/AbstractRegistrationOperation.java]: right. I'll fix that.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java [hunk 2]
```diff
@@ -622,13 +597,24 @@ public void handle(Packet packet) {
...
+    @Override
+    public Operation getPreJoinOperation() {
+        // pre-join operations are only sent by master member
+        return getRegistrationOperation();
+    }
+
+    @Override
+    public Operation getPostJoinOperation() {
+        // send post join operation only if this is the newly joining member
+        return nodeEngine.getClusterService().isMaster() ? null : getRegistrationOperation();
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java]: While operating in 3.8 cluster version, shouldn't the post-join op be delivered unconditionally?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java [hunk 3]
```diff
@@ -240,9 +240,9 @@ public void start() {
 
         metricsRegistry.collectMetrics(operationService, proxyService, eventService, operationParker);
 
+        operationService.start();
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: Can you elaborate why this change in service startup order was necessary?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java (hunk 2 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/eventservice/impl/EventServiceImpl.java]: you're right. I forgot 3.8 compat here completely.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java (hunk 3 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: Problem is, `DistributedScheduledExecutorService` registers partition lost listener in `init()` phase. When `invokeOnStableClusterSerial()` is called, it expects OperationService to be up. That's why I moved `operationService.start()` to above `serviceManager.start()`.

Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: I think I can solve this issue without changing service start order, by returning immediately from `invokeOnStableClusterSerial()` if node is not joined the cluster yet.

Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: Sounds right
